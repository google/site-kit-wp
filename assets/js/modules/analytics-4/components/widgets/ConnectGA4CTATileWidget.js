/**
 * ConnectGA4CTATileWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import ConnectModuleCTATile from '../../../../components/KeyMetrics/ConnectModuleCTATile';
import useWidgetStateEffect from '../../../../googlesitekit/widgets/hooks/useWidgetStateEffect';
import {
	CORE_USER,
	keyMetricsGA4Widgets,
} from '../../../../googlesitekit/datastore/user/constants';
import Null from '../../../../components/Null';

const metadata = { moduleSlug: 'analytics-4' };

export default function ConnectGA4CTATileWidget( { Widget, widgetSlug } ) {
	const ga4DependantKeyMetrics = useSelect( ( select ) => {
		const keyMetrics = select( CORE_USER ).getKeyMetrics();

		if ( ! keyMetrics ) {
			return [];
		}

		return keyMetrics.filter( ( keyMetric ) =>
			keyMetricsGA4Widgets.includes( keyMetric )
		).length;
	} );

	// Only render the CTA if there are 3 or fewer GA4-dependent key metrics tiles.
	// Otherwise, render `Null` to hide the CTA.
	// If there are four GA4-dependent key metrics tiles, the `ConnectGA4CTAWidget` will be rendered instead.
	const ConnectModuleCTAComponent =
		ga4DependantKeyMetrics > 3 ? Null : ConnectModuleCTATile;

	useWidgetStateEffect( widgetSlug, ConnectModuleCTAComponent, metadata );

	// Note that we need to render the `Widget` component as a wrapper here so this component will display
	// correctly when used as a `FallbackComponent` for the `whenActive` HOC. Conversely, when `ConnectModuleCTATile`
	// is rendered as an `OverrideComponent` in `WidgetRenderer` (as a result of multiple adjacent widgets rendering
	// this component and thus sharing the same state), it is wrapped with a `Widget` component - the net
	// result being the ConnectModuleCTATile is correctly wrapped in a `Widget` in both cases.
	return (
		<Widget>
			<ConnectModuleCTATile { ...metadata } />
		</Widget>
	);
}

ConnectGA4CTATileWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	widgetSlug: PropTypes.string.isRequired,
};
