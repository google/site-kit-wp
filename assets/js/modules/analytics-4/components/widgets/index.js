/**
 * GA4 Key Metrics widget components.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as ConnectGA4CTATileWidget } from './ConnectGA4CTATileWidget';
export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
export { default as EngagedTrafficSourceWidget } from './EngagedTrafficSourceWidget';
export { default as LeastEngagingPagesWidget } from './LeastEngagingPagesWidget';
export { default as MostEngagingPagesWidget } from './MostEngagingPagesWidget';
export { default as NewVisitorsWidget } from './NewVisitorsWidget';
export { default as PopularContentWidget } from './PopularContentWidget';
export { default as PopularProductsWidget } from './PopularProductsWidget';
export { default as ReturningVisitorsWidget } from './ReturningVisitorsWidget';
export { default as TopCitiesWidget } from './TopCitiesWidget';
export { default as TopCitiesDrivingLeadsWidget } from './TopCitiesDrivingLeadsWidget';
export { default as TopCitiesDrivingAddToCartWidget } from './TopCitiesDrivingAddToCartWidget';
export { default as TopCitiesDrivingPurchasesWidget } from './TopCitiesDrivingPurchasesWidget';
export { default as TopDeviceDrivingPurchasesWidget } from './TopDeviceDrivingPurchasesWidget';
export { default as TopCountriesWidget } from './TopCountriesWidget';
export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
export { default as TopTrafficSourceDrivingAddToCartWidget } from './TopTrafficSourceDrivingAddToCartWidget';
export { default as TopTrafficSourceDrivingLeadsWidget } from './TopTrafficSourceDrivingLeadsWidget';
export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
export { default as TopTrafficSourceDrivingPurchasesWidget } from './TopTrafficSourceDrivingPurchasesWidget';
export { default as PagesPerVisitWidget } from './PagesPerVisitWidget';
export { default as VisitLengthWidget } from './VisitLengthWidget';
export { default as TopReturningVisitorPages } from './TopReturningVisitorPages';
export { default as VisitsPerVisitorWidget } from './VisitsPerVisitorWidget';
export { default as TopRecentTrendingPagesWidget } from './TopRecentTrendingPagesWidget';
export { default as TopCategoriesWidget } from './TopCategoriesWidget';
export { default as TopPagesDrivingLeadsWidget } from './TopPagesDrivingLeadsWidget';
export { default as PopularAuthorsWidget } from './PopularAuthorsWidget';
