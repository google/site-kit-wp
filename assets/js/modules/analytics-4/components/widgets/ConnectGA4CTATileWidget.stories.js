/**
 * ConnectGA4CTATileWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import {
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	provideUserCapabilities,
} from '../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectAdSenseCTATile'
)( ConnectGA4CTATileWidget );

function Template() {
	return <WidgetWithComponentProps />;
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectGA4CTATileWidget';
Default.scenario = {
	label: 'KeyMetrics/ConnectGA4CTATileWidget',
	delay: 250,
};

export default {
	title: 'Key Metrics/ConnectGA4CTATileWidget',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				provideUserAuthentication( registry );
				provideUserCapabilities( registry );
				provideModules( registry, [
					{
						slug: 'analytics-4',
						active: false,
						connected: false,
					},
				] );
				provideModuleRegistrations( registry );
				provideSiteInfo( registry, {
					postTypes: [ { slug: 'product', label: 'Product' } ],
				} );
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
