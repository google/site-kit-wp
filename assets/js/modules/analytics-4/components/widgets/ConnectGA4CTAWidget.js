/**
 * ConnectGA4CTA component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useCallback, useEffect, useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { SpinnerButton } from 'googlesitekit-components';
import KeyMetricsCTAContent from '../../../../components/KeyMetrics/KeyMetricsCTAContent';
import KeyMetricsCTAFooter from '../../../../components/KeyMetrics/KeyMetricsCTAFooter';
import { CORE_MODULES } from '../../../../googlesitekit/modules/datastore/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { CORE_WIDGETS } from '../../../../googlesitekit/widgets/datastore/constants';
import { AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY } from '../../../../googlesitekit/widgets/default-areas';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';
import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import { KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY } from '../../constants';
import useActivateModuleCallback from '../../../../hooks/useActivateModuleCallback';
import useCompleteModuleActivationCallback from '../../../../hooks/useCompleteModuleActivationCallback';
import { useDebounce } from '../../../../hooks/useDebounce';

export default function ConnectGA4CTAWidget( { Widget, WidgetNull } ) {
	const ga4DependantKeyMetrics = useSelect( ( select ) => {
		const keyMetrics = select( CORE_USER ).getKeyMetrics();
		const widgets = select( CORE_WIDGETS ).getWidgets(
			AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY
		);

		if ( ! keyMetrics || ! widgets ) {
			return [];
		}

		return widgets.filter(
			( { slug, modules } ) =>
				keyMetrics.includes( slug ) && modules.includes( 'analytics-4' )
		);
	} );
	const isAnalyticsActive = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleActive( 'analytics-4' )
	);
	const isNavigatingToReauthURL = useSelect( ( select ) => {
		const adminReauthURL =
			select( MODULES_ANALYTICS_4 ).getAdminReauthURL();

		if ( ! adminReauthURL ) {
			return false;
		}

		return select( CORE_LOCATION ).isNavigatingTo( adminReauthURL );
	} );
	const isActivatingAnalytics = useSelect( ( select ) =>
		select( CORE_MODULES ).isFetchingSetModuleActivation(
			'analytics-4',
			true
		)
	);
	const connectGA4URL = useSelect( ( select ) => {
		const settingsURL = select( CORE_SITE ).getAdminURL(
			'googlesitekit-settings'
		);

		return `${ settingsURL }#connected-services/analytics-4/edit`;
	} );
	const isNavigatingToGA4URL = useSelect( ( select ) =>
		select( CORE_LOCATION ).isNavigatingTo( connectGA4URL )
	);

	const { dismissItem } = useDispatch( CORE_USER );

	const activateAnalytics = useActivateModuleCallback( 'analytics-4' );
	const completeAnalyticsActivation =
		useCompleteModuleActivationCallback( 'analytics-4' );

	const handleCTAClick = useCallback( () => {
		if ( isAnalyticsActive ) {
			return completeAnalyticsActivation();
		}

		activateAnalytics();
	}, [ activateAnalytics, completeAnalyticsActivation, isAnalyticsActive ] );

	const [ inProgress, setInProgress ] = useState( false );

	/*
	 * Using debounce here because the spinner has to render across two separate calls.
	 * Rather than risk it flickering on and off in between the activation call completing and
	 * the navigate call starting, we will just set a debounce to keep the spinner for 3 seconds.
	 */
	const debouncedSetInProgress = useDebounce( setInProgress, 3000 );

	useEffect( () => {
		if (
			isActivatingAnalytics ||
			isNavigatingToReauthURL ||
			isNavigatingToGA4URL
		) {
			setInProgress( true );
		} else {
			debouncedSetInProgress( false );
		}
	}, [
		isActivatingAnalytics,
		isNavigatingToReauthURL,
		debouncedSetInProgress,
		isNavigatingToGA4URL,
	] );

	const isDismissed = useSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed(
			KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY
		)
	);

	if ( isDismissed !== false || ga4DependantKeyMetrics.length < 4 ) {
		return <WidgetNull />;
	}

	return (
		<Widget
			noPadding
			Footer={ () => (
				<KeyMetricsCTAFooter
					onActionClick={ () =>
						dismissItem(
							KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY
						)
					}
					showDismiss
				/>
			) }
		>
			<KeyMetricsCTAContent
				className="googlesitekit-km-connect-ga4-cta"
				title={ __( 'Analytics is disconnected', 'google-site-kit' ) }
				description={ __(
					'Metrics cannot be displayed without Analytics',
					'google-site-kit'
				) }
				actions={
					<SpinnerButton
						onClick={ handleCTAClick }
						isSaving={ inProgress }
						disabled={ inProgress }
					>
						{ __( 'Connect Analytics', 'google-site-kit' ) }
					</SpinnerButton>
				}
			/>
		</Widget>
	);
}
