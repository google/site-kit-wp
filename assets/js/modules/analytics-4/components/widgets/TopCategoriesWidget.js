/**
 * TopCategoriesWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { compose } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { useSelect, useInViewSelect } from 'googlesitekit-data';
import {
	CORE_USER,
	KM_ANALYTICS_TOP_CATEGORIES,
} from '../../../../googlesitekit/datastore/user/constants';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { listFormat, numFmt } from '../../../../util';
import { ZeroDataMessage } from '../common';
import {
	MetricTileTable,
	MetricTileTablePlainText,
} from '../../../../components/KeyMetrics';
import whenActive from '../../../../util/when-active';
import withCustomDimensions from '../../utils/withCustomDimensions';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';
import { splitCategories } from '../../utils';

/**
 * Gets the report options for the Top Categories widget.
 *
 * @since 1.113.0
 *
 * @param {Function} select Data store 'select' function.
 * @return {Object} The report options.
 */
function getReportOptions( select ) {
	const dates = select( CORE_USER ).getDateRangeDates( {
		offsetDays: DATE_RANGE_OFFSET,
	} );

	return {
		...dates,
		dimensions: [ 'customEvent:googlesitekit_post_categories' ],
		dimensionFilters: {
			// Make sure that we select only rows without (not set) records.
			'customEvent:googlesitekit_post_categories': {
				filterType: 'emptyFilter',
				notExpression: true,
			},
		},
		metrics: [ { name: 'screenPageViews' } ],
		orderby: [
			{
				metric: {
					metricName: 'screenPageViews',
				},
				desc: true,
			},
		],
		limit: 3,
		keepEmptyRows: false,
	};
}

function TopCategoriesWidget( { Widget } ) {
	const topCategoriesReportOptions = useSelect( getReportOptions );

	const topCategoriesReport = useInViewSelect(
		( select ) =>
			select( MODULES_ANALYTICS_4 ).getReport(
				topCategoriesReportOptions
			),
		[ topCategoriesReportOptions ]
	);

	const error = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getErrorForSelector( 'getReport', [
			topCategoriesReportOptions,
		] )
	);

	const loading = useSelect(
		( select ) =>
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ topCategoriesReportOptions ]
			)
	);

	const { rows = [] } = topCategoriesReport || {};

	const columns = [
		{
			field: 'dimensionValues',
			Component( { fieldValue } ) {
				const [ categories ] = fieldValue;

				const categoriesList =
					typeof categories?.value === 'string'
						? splitCategories( categories.value )
						: [];

				const categoriesString = listFormat(
					// All values _must_ be a string or format will throw an error.
					categoriesList.map( String ),
					{ style: 'narrow' }
				);

				return (
					<MetricTileTablePlainText content={ categoriesString } />
				);
			},
		},
		{
			field: 'metricValues.0.value',
			Component( { fieldValue } ) {
				return <strong>{ numFmt( fieldValue ) }</strong>;
			},
		},
	];

	return (
		<MetricTileTable
			Widget={ Widget }
			widgetSlug={ KM_ANALYTICS_TOP_CATEGORIES }
			loading={ loading }
			rows={ rows }
			columns={ columns }
			ZeroState={ ZeroDataMessage }
			error={ error }
			moduleSlug="analytics-4"
		/>
	);
}

TopCategoriesWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};

export default compose(
	whenActive( {
		moduleName: 'analytics-4',
		FallbackComponent: ConnectGA4CTATileWidget,
	} ),
	withCustomDimensions( {
		reportOptions: getReportOptions,
	} )
)( TopCategoriesWidget );
