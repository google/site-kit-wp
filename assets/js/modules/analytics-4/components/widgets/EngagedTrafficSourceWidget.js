/**
 * EngagedTrafficSourceWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { get } from 'lodash';

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect, useInViewSelect } from 'googlesitekit-data';
import {
	CORE_USER,
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { MetricTileText } from '../../../../components/KeyMetrics';
import { numFmt } from '../../../../util';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';
import whenActive from '../../../../util/when-active';

function EngagedTrafficSourceWidget( props ) {
	const { Widget } = props;

	const dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
			compare: true,
		} )
	);

	const reportOptions = {
		...dates,
		dimensions: [ 'sessionDefaultChannelGroup' ],
		metrics: [ { name: 'engagedSessions' } ],
		orderBy: 'engagedSessions',
		limit: 1,
	};

	const report = useInViewSelect(
		( select ) => select( MODULES_ANALYTICS_4 ).getReport( reportOptions ),
		[ reportOptions ]
	);

	const error = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getErrorForSelector( 'getReport', [
			reportOptions,
		] )
	);

	const loading = useSelect(
		( select ) =>
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ reportOptions ]
			)
	);

	const { rows = [], totals = [] } = report || {};
	const makeFilter = ( dateRange, dimensionIndex ) => ( row ) =>
		get( row, `dimensionValues.${ dimensionIndex }.value` ) === dateRange;

	const topTrafficSource =
		rows.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
			?.dimensionValues?.[ 0 ].value || '-';

	const currentEngagedSessions =
		parseFloat(
			rows.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const currentTotalEngagedSessions =
		parseFloat(
			totals.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const currentEngagedSessionsRate =
		currentEngagedSessions / currentTotalEngagedSessions || 0;

	const previousEngagedSessions =
		parseFloat(
			rows.filter( makeFilter( 'date_range_1', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const previousTotalEngagedSessions =
		parseFloat(
			totals.filter( makeFilter( 'date_range_1', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const previousEngagedSessionsRate =
		previousEngagedSessions / previousTotalEngagedSessions || 0;

	const format = {
		style: 'percent',
		signDisplay: 'never',
		maximumFractionDigits: 1,
	};

	return (
		<MetricTileText
			Widget={ Widget }
			widgetSlug={ KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE }
			metricValue={ topTrafficSource }
			metricValueFormat={ format }
			subText={ sprintf(
				/* translators: 1. Percentage of total engaged sessions. 2: Total number of engaged sessions. */
				__( '%1$s of %2$s engaged sessions', 'google-site-kit' ),
				numFmt( currentEngagedSessionsRate, format ),
				numFmt( currentTotalEngagedSessions, { style: 'decimal' } )
			) }
			previousValue={ previousEngagedSessionsRate }
			currentValue={ currentEngagedSessionsRate }
			loading={ loading }
			error={ error }
			moduleSlug="analytics-4"
		/>
	);
}

EngagedTrafficSourceWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};

export default whenActive( {
	moduleName: 'analytics-4',
	FallbackComponent: ConnectGA4CTATileWidget,
} )( EngagedTrafficSourceWidget );
