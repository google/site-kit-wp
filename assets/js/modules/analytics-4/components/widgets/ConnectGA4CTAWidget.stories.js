/**
 * ConnectGA4CTAWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
	KM_ANALYTICS_RETURNING_VISITORS,
	KM_ANALYTICS_NEW_VISITORS,
	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import { provideKeyMetrics } from '../../../../../../tests/js/utils';
import { provideKeyMetricsWidgetRegistrations } from '../../../../components/KeyMetrics/test-utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import ConnectGA4CTAWidget from './ConnectGA4CTAWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectGA4CTA'
)( ConnectGA4CTAWidget );

function Template() {
	return (
		<div className="googlesitekit-widget-area--mainDashboardKeyMetricsPrimary">
			<div className="googlesitekit-widget--keyMetricsConnectGA4All">
				<WidgetWithComponentProps />
			</div>
		</div>
	);
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectGA4CTAWidget';
Default.scenario = {
	label: 'KeyMetrics/ConnectGA4CTAWidget',
};

export default {
	title: 'Key Metrics/ConnectGA4CTAWidget',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				const keyMetricWidgets = [
					KM_ANALYTICS_RETURNING_VISITORS,
					KM_ANALYTICS_NEW_VISITORS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
					KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				];

				provideKeyMetrics( registry, {
					widgetSlugs: keyMetricWidgets,
				} );

				provideKeyMetricsWidgetRegistrations(
					registry,
					keyMetricWidgets.reduce(
						( acc, widget ) => ( {
							...acc,
							[ widget ]: { modules: [ 'analytics-4' ] },
						} ),
						{}
					)
				);
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
