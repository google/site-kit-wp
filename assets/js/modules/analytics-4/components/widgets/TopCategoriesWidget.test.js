/**
 * TopCategoriesWidget component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { render } from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	provideKeyMetrics,
	provideModules,
	provideUserAuthentication,
} from '../../../../../../tests/js/utils';
import { provideAnalytics4MockReport } from '../../utils/data-mock';
import { getWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import {
	CORE_USER,
	KM_ANALYTICS_TOP_CATEGORIES,
	KM_ANALYTICS_TOP_CITIES,
} from '../../../../googlesitekit/datastore/user/constants';
import { KEY_METRICS_WIDGETS } from '../../../../components/KeyMetrics/key-metrics-widgets';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import TopCategoriesWidget from './TopCategoriesWidget';

describe( 'TopCategoriesWidget', () => {
	const { Widget } = getWidgetComponentProps( KM_ANALYTICS_TOP_CITIES );

	let registry;
	beforeEach( () => {
		registry = createTestRegistry();
		registry.dispatch( CORE_USER ).setReferenceDate( '2020-09-08' );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			availableCustomDimensions:
				KEY_METRICS_WIDGETS[ KM_ANALYTICS_TOP_CATEGORIES ]
					.requiredCustomDimensions,
		} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveIsGatheringData( false );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveIsCustomDimensionGatheringData(
				KEY_METRICS_WIDGETS[ KM_ANALYTICS_TOP_CATEGORIES ]
					.requiredCustomDimensions[ 0 ],
				false
			);

		provideUserAuthentication( registry );
		provideModules( registry, [
			{
				slug: 'analytics-4',
				active: true,
				connected: true,
			},
		] );
		provideKeyMetrics( registry );
	} );

	it( 'renders correctly with the expected metrics', async () => {
		provideAnalytics4MockReport( registry, {
			...registry.select( CORE_USER ).getDateRangeDates( {
				offsetDays: DATE_RANGE_OFFSET,
			} ),
			dimensions: [ 'customEvent:googlesitekit_post_categories' ],
			dimensionFilters: {
				'customEvent:googlesitekit_post_categories': {
					filterType: 'stringFilter',
					matchType: 'EXACT',
					value: '(not set)',
					notExpression: true,
				},
			},
			metrics: [ { name: 'screenPageViews' } ],
			orderby: [
				{
					metric: {
						metricName: 'screenPageViews',
					},
					desc: true,
				},
			],
			limit: 3,
		} );

		const { container, waitForRegistry } = render(
			<TopCategoriesWidget
				widgetSlug={ KM_ANALYTICS_TOP_CATEGORIES }
				Widget={ Widget }
			/>,
			{ registry }
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );
} );
