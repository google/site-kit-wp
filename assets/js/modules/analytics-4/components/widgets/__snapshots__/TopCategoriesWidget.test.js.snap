// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopCategoriesWidget renders correctly with the expected metrics 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsTopCities googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--table"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-km-widget-tile__title"
          >
            Top categories by pageviews
          </h3>
          <span
            class="googlesitekit-info-tooltip"
            title="Categories that your site visitors viewed the most"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__table"
          >
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <p
                  class="googlesitekit-km-widget-tile__table-plain-text"
                >
                  Entertainment, Sports, Media
                </p>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  62
                </strong>
              </div>
            </div>
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <p
                  class="googlesitekit-km-widget-tile__table-plain-text"
                >
                  Wealth
                </p>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  35
                </strong>
              </div>
            </div>
            <div
              class="googlesitekit-table__body-row"
            >
              <div
                class="googlesitekit-table__body-item"
              >
                <p
                  class="googlesitekit-km-widget-tile__table-plain-text"
                >
                  Health
                </p>
              </div>
              <div
                class="googlesitekit-table__body-item"
              >
                <strong>
                  21
                </strong>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
